easybuild/easyblocks/l/lammps.py

```diff
@@ -211,6 +224,23 @@ def configure_step(self, **kwargs):
             if '-DFFT_PACK=' not in self.cfg['configopts']:
                 self.cfg.update('configopts', '-DFFT_PACK=array')
 
+        # Additional linker flags
+        lammps_link_libs = []
+
+        # Atomistica
+        if get_software_root('Atomistica'):
+            lammps_link_libs += ['-latomistica']
+            if self.toolchain.options.get('usempi', None):
+                if self.toolchain.mpi_family() == toolchain.OPENMPI:
+                    lammps_link_libs += ['-lmpi_usempif08', '-lmpi_usempi_ignore_tkr', '-lmpi_mpifh', '-lmpi']
+                elif self.toolchain.mpi_family() == toolchain.INTELMPI:
+                    lammps_link_libs += ['-lmpifort', '-lmpi', '-lmpigi']
```

Comment:
This feels wrong. The MPI compiler wrappers should add these, and especially for OpenMPI the names of the libraries depend on how OpenMPI was built.

Reply:
I share the feeling. However, even using the MPI compiler wrappers hits `undefined reference` errors to MPI symbols at the very last link stage

Full log at https://gist.github.com/lexming/40d78c44817dd7def3c6a260351fea92
>[100%] Linking CXX executable lmp
/theia/home/apps/CO7/ivybridge-ib/software/CMake/3.15.3-GCCcore-8.3.0/bin/cmake -E cmake_link_script CMakeFiles/lmp.dir/link.txt --verbose=1
/apps/brussel/CO7/ivybridge-ib/software/OpenMPI/3.1.4-GCC-8.3.0/bin/mpicxx  -O2 -ftree-vectorize -march=native -fno-math-errno -fopenmp -std=c++11 -fopenmp -O3 -DNDEBUG  -L/user/brussel/101/vsc10122/.local/easybuild-ivybridge/software/Atomistica/0.10.1-foss-2019b-Python-3.7.4/lib -L/apps/brussel/CO7/ivybridge-ib/software/PLUMED/2.5.3-foss-2019b-Python-3.7.4/lib -L/apps/brussel/CO7/ivybridge-ib/software/yaff/1.6.0-foss-2019b-Python-3.7.4/lib -L/apps/brussel/CO7/ivybridge-ib/software/kim-api/2.1.3-foss-2019b/lib64 -L/apps/brussel/CO7/ivybridge-ib/software/FFTW/3.3.8-gompi-2019b/lib -L/apps/brussel/CO7/ivybridge-ib/software/Voro++/0.4.6-foss-2019b/lib -L/apps/brussel/CO7/ivybridge-ib/software/FFmpeg/4.2.1-GCCcore-8.3.0/lib -L/apps/brussel/CO7/ivybridge-ib/software/libxml2/2.9.9-GCCcore-8.3.0/lib -L/apps/brussel/CO7/ivybridge-ib/software/PCRE/8.43-GCCcore-8.3.0/lib -L/apps/brussel/CO7/ivybridge-ib/software/HDF5/1.10.5-gompi-2019b/lib -L/apps/brussel/CO7/ivybridge-ib/software/zlib/1.2.11-GCCcore-8.3.0/lib -L/apps/brussel/CO7/ivybridge-ib/software/GSL/2.6-GCC-8.3.0/lib -L/apps/brussel/CO7/ivybridge-ib/software/netCDF/4.7.1-gompi-2019b/lib64 -L/apps/brussel/CO7/ivybridge-ib/software/libjpeg-turbo/2.0.3-GCCcore-8.3.0/lib64 -L/apps/brussel/CO7/ivybridge-ib/software/libpng/1.6.37-GCCcore-8.3.0/lib -L/apps/brussel/CO7/ivybridge-ib/software/Python/3.7.4-GCCcore-8.3.0/lib -L/apps/brussel/CO7/ivybridge-ib/software/archspec/0.1.0-GCCcore-8.3.0-Python-3.7.4/lib -L/apps/brussel/CO7/ivybridge-ib/software/ScaLAPACK/2.0.2-gompi-2019b/lib -L/apps/brussel/CO7/ivybridge-ib/software/OpenBLAS/0.3.7-GCC-8.3.0/lib -L/apps/brussel/CO7/ivybridge-ib/software/GCCcore/8.3.0/lib64 -L/apps/brussel/CO7/ivybridge-ib/software/GCCcore/8.3.0/lib -rdynamic CMakeFiles/lmp.dir/theia/scratch/brussel/101/vsc10122/easybuild-ivb/build/LAMMPS/7Aug2019/foss-2019b-Python-3.7.4-kokkos-atomistica/lammps-stable_7Aug2019/src/main.cpp.o  -o lmp -Wl,-rpath,/theia/scratch/brussel/101/vsc10122/easybuild-ivb/build/LAMMPS/7Aug2019/foss-2019b-Python-3.7.4-kokkos-atomistica/easybuild_obj:/theia/scratch/brussel/101/vsc10122/easybuild-ivb/build/LAMMPS/7Aug2019/foss-2019b-Python-3.7.4-kokkos-atomistica/easybuild_obj/lib/kokkos: liblammps.so.0 -latomistica -ljpeg -lpng -lfftw3 -lpython3.7m -Wl,-Bstatic -lvoro++ -lplumed -Wl,-Bdynamic -ldl -lstdc++ -lfftw3 -lgsl -lz -lopenblas -lgfortran -lm -lpthread -L/apps/brussel/CO7/ivybridge-ib/software/SciPy-bundle/2019.10-foss-2019b-Python-3.7.4/lib -L/apps/brussel/CO7/ivybridge-ib/software/Python/3.7.4-GCCcore-8.3.0/lib -L/apps/brussel/CO7/ivybridge-ib/software/libmatheval/1.1.11-GCCcore-8.3.0/lib -L/apps/brussel/CO7/ivybridge-ib/software/GSL/2.6-GCC-8.3.0/lib -L/apps/brussel/CO7/ivybridge-ib/software/zlib/1.2.11-GCCcore-8.3.0/lib -L/apps/brussel/CO7/ivybridge-ib/software/FFTW/3.3.8-gompi-2019b/lib -L/apps/brussel/CO7/ivybridge-ib/software/ScaLAPACK/2.0.2-gompi-2019b/lib -L/apps/brussel/CO7/ivybridge-ib/software/OpenBLAS/0.3.7-GCC-8.3.0/lib -L/apps/brussel/CO7/ivybridge-ib/software/GCCcore/8.3.0/lib64 -L/apps/brussel/CO7/ivybridge-ib/software/GCCcore/8.3.0/lib -rdynamic -Wl,-Bsymbolic -ldl -lnetcdf -lcurl -L/apps/brussel/CO7/ivybridge-ib/software/kim-api/2.1.3-foss-2019b/lib64 -Wl,-rpath,/apps/brussel/CO7/ivybridge-ib/software/kim-api/2.1.3-foss-2019b/lib64 -lkim-api libcsmpi.a -lm libpoems.a libatc.a -lopenblas libcolvars.a libh5md.a -lhdf5 -lsz -lz -lm -lpthread lib/kokkos/libkokkos.so -fopenmp -ldl -lrt -ltbbmalloc 
liblammps.so.0: error: undefined reference to 'mpi_comm_set_errhandler_'
liblammps.so.0: error: undefined reference to 'mpi_cart_create_'
liblammps.so.0: error: undefined reference to 'mpi_finalize_'
liblammps.so.0: error: undefined reference to 'mpi_comm_split_'
liblammps.so.0: error: undefined reference to 'mpi_init_'
liblammps.so.0: error: undefined reference to 'mpi_cart_coords_'
liblammps.so.0: error: undefined reference to 'mpi_initialized_'
liblammps.so.0: error: undefined reference to 'mpi_comm_free_'
liblammps.so.0: error: undefined reference to 'mpi_allreduce_'
liblammps.so.0: error: undefined reference to 'mpi_scan_'
liblammps.so.0: error: undefined reference to 'mpi_bcast_'
liblammps.so.0: error: undefined reference to 'mpi_allgatherv_'
liblammps.so.0: error: undefined reference to 'mpi_allgather_'
liblammps.so.0: error: undefined reference to 'mpi_barrier_'
liblammps.so.0: error: undefined reference to 'mpi_cart_shift_'
liblammps.so.0: error: undefined reference to 'mpi_get_count_'
liblammps.so.0: error: undefined reference to 'mpi_sendrecv_'
liblammps.so.0: error: undefined reference to 'mpi_error_string_'
liblammps.so.0: error: undefined reference to 'mpi_comm_rank_'
liblammps.so.0: error: undefined reference to 'mpi_comm_size_'
liblammps.so.0: error: undefined reference to 'mpi_abort_'
collect2: error: ld returned 1 exit status

I do not know if there is a better way to fix this link errors. I'll check if the list of flags can be stripped down.

Reply:
Was liblammps.so created with the mpi wrappers? My guess is it wasn't, it needs to be.

Reply:
`liblammps.so` was created with `mpicxx` and had no errors or warnings...
>[100%] Linking CXX shared library liblammps.so
/theia/home/apps/CO7/ivybridge-ib/software/CMake/3.15.3-GCCcore-8.3.0/bin/cmake -E cmake_link_script CMakeFiles/lammps.dir/link.txt --verbose=1
/apps/brussel/CO7/ivybridge-ib/software/OpenMPI/3.1.4-GCC-8.3.0/bin/mpicxx -fPIC -O2 -ftree-vectorize -march=native -fno-math-errno -fopenmp -std=c++11 -fopenmp -O3 -DNDEBUG -L/user/brussel/101/vsc10122/.local/easybuild-ivybridge/software/Atomistica/0.10.1-foss-2019b-Python-3.7.4/lib -L/apps/brussel/CO7/ivybridge-ib/software/PLUMED/2.5.3-foss-2019b-Python-3.7.4/lib -L/apps/brussel/CO7/ivybridge-ib/software/yaff/1.6.0-foss-2019b-Python-3.7.4/lib -L/apps/brussel/CO7/ivybridge-ib/software/kim-api/2.1.3-foss-2019b/lib64 -L/apps/brussel/CO7/ivybridge-ib/software/FFTW/3.3.8-gompi-2019b/lib -L/apps/brussel/CO7/ivybridge-ib/software/Voro++/0.4.6-foss-2019b/lib -L/apps/brussel/CO7/ivybridge-ib/software/FFmpeg/4.2.1-GCCcore-8.3.0/lib -L/apps/brussel/CO7/ivybridge-ib/software/libxml2/2.9.9-GCCcore-8.3.0/lib -L/apps/brussel/CO7/ivybridge-ib/software/PCRE/8.43-GCCcore-8.3.0/lib -L/apps/brussel/CO7/ivybridge-ib/software/HDF5/1.10.5-gompi-2019b/lib -L/apps/brussel/CO7/ivybridge-ib/software/zlib/1.2.11-GCCcore-8.3.0/lib -L/apps/brussel/CO7/ivybridge-ib/software/GSL/2.6-GCC-8.3.0/lib -L/apps/brussel/CO7/ivybridge-ib/software/netCDF/4.7.1-gompi-2019b/lib64 -L/apps/brussel/CO7/ivybridge-ib/software/libjpeg-turbo/2.0.3-GCCcore-8.3.0/lib64 -L/apps/brussel/CO7/ivybridge-ib/software/libpng/1.6.37-GCCcore-8.3.0/lib -L/apps/brussel/CO7/ivybridge-ib/software/Python/3.7.4-GCCcore-8.3.0/lib -L/apps/brussel/CO7/ivybridge-ib/software/archspec/0.1.0-GCCcore-8.3.0-Python-3.7.4/lib -L/apps/brussel/CO7/ivybridge-ib/software/ScaLAPACK/2.0.2-gompi-2019b/lib -L/apps/brussel/CO7/ivybridge-ib/software/OpenBLAS/0.3.7-GCC-8.3.0/lib -L/apps/brussel/CO7/ivybridge-ib/software/GCCcore/8.3.0/lib64 -L/apps/brussel/CO7/ivybridge-ib/software/GCCcore/8.3.0/lib -shared -Wl,-soname,liblammps.so.0 -o liblammps.so.0 @CMakeFiles/lammps.dir/objects1.rsp @CMakeFiles/lammps.dir/objects2.rsp @CMakeFiles/lammps.dir/objects3.rsp -Wl,-rpath,/theia/scratch/brussel/101/vsc10122/easybuild-ivb/build/LAMMPS/7Aug2019/foss-2019b-Python-3.7.4-kokkos-atomistica/easybuild_obj/lib/kokkos: -latomistica -ljpeg -lpng -lz -lfftw3 -lpython3.7m -Wl,-Bstatic -lvoro++ -lplumed -Wl,-Bdynamic -ldl -lstdc++ -lfftw3 -lgsl -lz -lopenblas -lgfortran -lm -lpthread -L/apps/brussel/CO7/ivybridge-ib/software/SciPy-bundle/2019.10-foss-2019b-Python-3.7.4/lib -L/apps/brussel/CO7/ivybridge-ib/software/Python/3.7.4-GCCcore-8.3.0/lib -L/apps/brussel/CO7/ivybridge-ib/software/libmatheval/1.1.11-GCCcore-8.3.0/lib -L/apps/brussel/CO7/ivybridge-ib/software/GSL/2.6-GCC-8.3.0/lib -L/apps/brussel/CO7/ivybridge-ib/software/zlib/1.2.11-GCCcore-8.3.0/lib -L/apps/brussel/CO7/ivybridge-ib/software/FFTW/3.3.8-gompi-2019b/lib -L/apps/brussel/CO7/ivybridge-ib/software/ScaLAPACK/2.0.2-gompi-2019b/lib -L/apps/brussel/CO7/ivybridge-ib/software/OpenBLAS/0.3.7-GCC-8.3.0/lib -L/apps/brussel/CO7/ivybridge-ib/software/GCCcore/8.3.0/lib64 -L/apps/brussel/CO7/ivybridge-ib/software/GCCcore/8.3.0/lib -rdynamic -Wl,-Bsymbolic -fopenmp -lnetcdf -lcurl -L/apps/brussel/CO7/ivybridge-ib/software/kim-api/2.1.3-foss-2019b/lib64 -Wl,-rpath,/apps/brussel/CO7/ivybridge-ib/software/kim-api/2.1.3-foss-2019b/lib64 -lkim-api libcsmpi.a -lm libpoems.a libatc.a libcolvars.a libh5md.a lib/kokkos/libkokkos.so -ltbbmalloc -ldl -lopenblas -lhdf5 -lsz -lz -lm -lpthread -fopenmp -ldl -lrt 
/theia/home/apps/CO7/ivybridge-ib/software/CMake/3.15.3-GCCcore-8.3.0/bin/cmake -E cmake_symlink_library liblammps.so.0 liblammps.so.0 liblammps.so

Reply:
@akesandgren I think that the issue is that `liblammps` is made with `mpicxx` and `libatomistica` with `mpifort`. Then when both are linked together `liblammps` is missing all fortran symbols from MPI. The only flags needed (and that I manually added to the easyblock) are MPI libs in fortran.

Reply:
I found https://github.com/LLNL/scr/pull/132 and that was added because dependencies on deprecated C++ bindings were introduced somehow (see issue https://github.com/LLNL/scr/issues/130). Maybe a similar looking patch might do the job?

Reply:
Ah.
The "correct" solution for OpenMPI is then to use
"mpifort --showme:libs" to find out which libs it needs instead of hard coding it.
Unfortunately there is no such option for mpiifort

Reply:
For IMPI you should also only need -lmpifort the rest should be added bu mpiicpc

Reply:
The ideal solution is perform the link step with the MPI Fortran compiler instead of the CXX one. Then all the fortran bindings are automatically added and it does not depend on the toolchain in use. The problem is that this needs adding a `LINKER_LANGUAGE` variable to a specific  `set_target_properties` in CMakeLists.txt.

Reply:
Setting `LINKER_LANGUAGE` is not easy to apply because it is not enough to link the final `lmp` executable with `mpifort`, other shared libraries also have references to MPI fortran bindings such as `liblammps.so`. So there is no simple command that can be done from this easyblock to fix the build of LAMMPS with Atomistica. The solution is to patch `CMakeLists.txt`.